'scala' command line tool takes only one source file as a parameter

'scala' command line tool is very useful as a linux script executor. Unfortunately it supports only one file which, can be passed as argument. But I think that in more advance system scripting it is very common to want to refactor part of scripts into separate source unit. For example I have few scripts in bash for backing up my system, but calculating paths is common part and is in separate file mirror-config.sh.

As I said before I can not do it easily using Scala as I can not provide more than one scala source file.

My current (imperfect) solution is as follows:

scala-utils.scala -> contains some common utilities for scripting
test.sh -> main script in scala wrapped in bash

test.sh file:
---------------------------------------------------
#!/bin/bash
source /root/bin/boot-scala.sh
!#

import java.io.File
import utils.Timer

object App {
  def main(args: Array[String]): Unit = {
    val timer = new Timer

    timer.start
    println("Executing '%s' in directory '%s'".format(args(1), args(0)))
    println("There are %d arguments passed to script.".format(args.length))
    println(args.mkString(" "))

    println("Time elapsed: " + timer.elapsed)
  }
}
---------------------------------------------------

boot-scala file:
---------------------------------------------------
#!/bin/sh
SCRIPT="$(cd "${0%/*}" 2>/dev/null; echo "$PWD"/"${0##*/}")"
DIR=`dirname "${SCRIPT}"}`
CP="/tmp/scala-cp/"
if [ ! -d $CP ]; then
  mkdir /tmp/scala-cp
  scalac -d $CP $DIR/*.scala
fi
exec scala -savecompiled -classpath $CP $0 $DIR $SCRIPT $@
---------------------------------------------------

As you can see in boot-scala.sh, I am extracting SCRIPT name and DIR name and then these values are passed to scala.
Problem is that I can only provide classpath to scala tool, so I have to compile manually my scala-utils.scala file and put classes into arbitrary /tmp/scala-cp.

If scala tool could have another command line argument e.g. -sources, then I could provide other files directly without making such a strange hacks.

Also it would be nice if compiled *.jar files could be stored in some internal scala directory, so they are not cluttering my scripts directory.

This is a use case for `-i`, but it doesn't seem to work:

```
$ scala -howtorun:script -i t4033b.scala -- t4033.scala
t4033.scala:17: error: not found: value ng
    println("ng:" + ng.klass)
                    ^
```
Normally, there's no way to know this, but `-i` says take these "init" files (terminated by double dash), for REPL, but you can force the script runner.

--- a/src/compiler/scala/tools/nsc/Global.scala
+++ b/src/compiler/scala/tools/nsc/Global.scala
@@ -38,7 +38,7 @@ import scala.tools.nsc.symtab.{Flags, SymbolTable, SymbolTrackers}
 import scala.tools.nsc.transform._
 import scala.tools.nsc.transform.patmat.PatternMatching
 import scala.tools.nsc.typechecker._
-import scala.tools.nsc.util.{ClassPath, returning}
+import scala.tools.nsc.util.ClassPath
 
 class Global(var currentSettings: Settings, reporter0: Reporter)
     extends SymbolTable
@@ -1604,8 +1604,11 @@ class Global(var currentSettings: Settings, reporter0: Reporter)
         val snap = profiler.beforePhase(Global.InitPhase)
 
         val sources: List[SourceFile] =
-          if (settings.script.isSetByUser && filenames.size > 1) returning(Nil)(_ => globalError("can only compile one script at a time"))
-          else filenames map getSourceFile
+          if (settings.script.isSetByUser && filenames.size > 1) {
+            globalError("can only compile one script at a time")
+            Nil
+          }
+          else filenames.map(getSourceFile)
 
         profiler.afterPhase(Global.InitPhase, snap)
         compileSources(sources)

--- a/src/compiler/scala/tools/nsc/ScriptRunner.scala
+++ b/src/compiler/scala/tools/nsc/ScriptRunner.scala
@@ -57,12 +57,12 @@ trait ScriptRunner {
 
 class DefaultScriptRunner(settings: GenericRunnerSettings) extends AbstractScriptRunner(settings) {
   protected def doCompile(scriptFile: String) = {
-    // Setting settings.script.value informs the compiler this is not a self-contained compilation unit.
-    settings.script.value = mainClass
     val reporter = new ConsoleReporter(settings)
     val compiler = newGlobal(settings, reporter)
-    val run      = new compiler.Run
-    run.compile(List(scriptFile))
+    if (settings.pastefiles.value.nonEmpty) new compiler.Run().compile(settings.pastefiles.value)
+    // Setting settings.script.value informs the compiler this is not a self-contained compilation unit.
+    settings.script.value = mainClass
+    new compiler.Run().compile(scriptFile :: Nil)
     !reporter.hasErrors
   }
 

